Feature: support ProcessPoolExecutor (Python >= 3.3 required)

asfaltboy opened the pull request:
## Change Summary

* include tests, some of which are skipped if on unsupported python version
* use `functools.partial` instead of directly wrapping the `request` method,
  regardless of executor used - avoid code branching
* use `functools.partial(Session.request, self)` instead of
  `super(FuturesSession, self).request` for pickling, again no code branching

## Notes / Limitations

* **Python version 3.4** (actually >= 3.3.5) is required in order for us to pickle methods, [due to python issue 9276](https://bugs.python.org/issue9276). Though we may be able to [use `copy_reg` to add this support to Python 2](http://stackoverflow.com/a/25161919/484127).
* **Python 3.5** is required to "create a session on demand", with lower versions an existing session is required. Probably due to one of the numerous `pickle` related issues [fixed in 3.5](https://docs.python.org/3/whatsnew/changelog.html).
* Adding custom attributes to the response object in the callback is supported in `ProcessPoolExecutor` but only when the callback also modifies `__attrs__` so it gets pickled and sent to parent process (see an example for that in the tests).
* As mentioned, the `background_callback` MUST be a top-level (or otherwise importable) object. Additionally, since an instance of `FuturesSession` is passed to the callback (as `self`), it too must be importable, otherwise `pickle` will complain.

## Conclusions

This relatively simple implementation fixes issue #11, however due to numerous caveats, I'd suggest we recommend it be used in Python 3.5 and above and explain the limitation of earlier versions. If earlier versions (like Python 2) are really required by someone they may need to implement a manual fix to the `pickle` implementation (e.g using `copy_reg`).

Personally, this PR solves my issue but I have yet to test it on production, hopefully, will do so in the near future.

ross commented on test_requests_futures.py:
```diff
@@ -4,10 +4,12 @@
-from requests import Response, session
 from os import environ
+import unittest
+import sys
```
Stylistically I prefer everything to be `from X import Y, Z` rather than `import X`... `X.whatever`

ross commented on requests_futures/sessions.py:
```diff
@@ -33,9 +40,12 @@ def __init__(self, executor=None, max_workers=2, session=None, *args,
-        * ProcessPoolExecutor is not supported b/c Response objects are
-          not picklable.
+          Additionally, if on Python < 3.5 an existing session object MUST be
+          passed when initializing with `ProcessPoolExecutor`.
```
Seems like this should either be a `*` or merged in with the point above.

ross replied:
What will the error message/failure case be if someone fails to do so? If it's non-obvious it'd be good to check for that case and throw a better one.

ross commented:
Sorry for the slow response. Super busy at work as of late. Looks pretty good. Few small comments inline.

asfaltboy commented:
No problemo. I added an exception, some documentation and updated tests. Unfortunately, the `README.rst` does not render so well in GitHub, so feel free to style the doc to your liking.

ross commented:
Thanks.
